### ISSUE TITLE
[Editor] text search is slow for large files

### ISSUE THREAD

**Issue by jasonLaster** [devtools-html/debugger.html#2355]
We currently use codemirror to find all of the search results in a document when we start searching. Unfortunately, this is a pretty slow search that we could speed up pretty easily.

Here is what we do [here](https://github.com/devtools-html/debugger.html/blob/master/src/utils/editor/source-search.js#L37-L40)

```js
  state.results = [];
  while (cursor.findNext()) {
    state.results.push(cursor.pos);
  }
```

Here is the data structure that it generates.
```json
{
  "from": {
    "line": 15,
    "ch": 16,
    "sticky": null
  },
  "to": {
    "line": 15,
    "ch": 19,
    "sticky": null
  },
  "match": [
  {
    0: "ati"
    input: "getOriginalLocation, getGeneratedLocation, isOriginalId"
    index: 16
  }
  ]
}
```

We could probably speed it up by getting a count of matches and then using codemirror to find the next or previous result on demand.

**Comment by wldcordeiro** [comment]
This is a good idea, we kind of do this already too with countMatches.

**Comment by ryanjduffy** [comment]
Would be best if you could defer the line calculation until it was needed to scroll it into view. Otherwise, you have to no only find the string but also derive the line number. You could incrementally sort it out (e.g. count the line for the first match, cache it, count the lines from the first match to the second match and sum it to the cached value, etc) but it's likely a lot of wasted effort since in most cases the user doesn't need every match.

**Comment by jasonLaster** [comment]
Yeah. That's a good idea. We probably can.

**Comment by wldcordeiro** [comment]
I didn't get a chance to test it yet but it might be possible to do something kind of like this to achieve what you suggested @jasonLaster not sure how it performs versus what @ryanjduffy suggested.

```js
sourceText
  .split(/\\r?\\n/)
  .reduce((acc, line) => {
    acc.push(query // query being the regexp object we return from buildQuery
      .exec(line)
      .map(...)); // format like codemirror's returns.
    return acc;
  });
```

**Comment by ryanjduffy** [comment]
I started on this using a variation of @wldcordeiro's suggestion. Initial search seemed to be faster with my rough algorithm but I was surprised when my naive 'only find the first instance in the line' approach succeeded in finding every instance in the line. Since that clearly couldn't have been the case, I discovered that the loop code in `getSearchState()` seems to be unnecessary as the incremental search in `doSearch()` and `searchNext` work fine without it.

**Comment by wldcordeiro** [comment]
@ryanjduffy Do you have a public branch or a patch you could share? I could test it out too.  😄

**Comment by ryanjduffy** [comment]
I don't have much to share. I'm not even convinced my search logic is valid since it seems to be superseded by the code in `doSearch()`. You can test my theory by commenting out lines 36-43 in `source-search.js`.

**Comment by gregtatum** [comment]
I just spent 60 seconds with the main thread hung waiting on a search for a minified file (beach ball unresponsive browser). Worse yet, I wasn't actively searching, I had a search already in the search box from a previous file. This feels pretty bad behavior going out on a major release :-/

The time is being spent in CodeMirror token matching: https://perfht.ml/2r0hOOK
Caused primarily by the garbage collection of the tokens: https://perfht.ml/2r0046c

STR:
* Visit: http://alteredqualia.com/xg/examples/hannibal_draco.html
* Open xg.min.625.js in the debugger.
* Search for `dofEnabled`

**Comment by jasonLaster** [comment]
Thanks @ryanjduffy you're absolutely right. I just made this PR https://github.com/devtools-html/debugger.html/pull/3008 to remove the search results, which speeds it up tremendously.

Now we just need to do something like wellington's algorithm to compute the location of the matches per document and then use that to make a new search.results list that we can use.

**Comment by jasonLaster** [comment]
Moved text search to a worker, which makes it smooth. We still need to get the match index though...

**Comment by jasonLaster** [comment]
closing in favor of a more specific issue #3204
